types/wordpress__block-editor/components/use-block-props.d.ts

```diff
@@ -0,0 +1,29 @@
+import { Ref } from 'react';
+
+export interface Reserved {
+    id: string;
+    role: 'document';
+    tabIndex: 0;
+    'aria-label': string;
+    'data-block': string;
+    'data-type': string;
+    'data-title': string;
+}
+
+export interface Merged {
+    className: string;
+    style: Record<string, unknown>;
+    ref: (value: unknown) => void;
+}
+
+export interface UseBlockProps {
+    <Props extends Record<string, unknown>>(
+        props?: Props & {
+            [K in keyof Props]: K extends keyof Reserved ? never : Props[K];
+        } & { ref?: Ref<unknown> },
```

Comment:
constraining to `Ref<unknown>` is a nice addition. We could also [thread the type through with a type parameter](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nASigZnAXzlyghDgHIoUBDAYxgoG4BYAKHeADsYUpd6aHAGc+ANxQATDHFIAbFAC5KkiHQCuIFDwqFWHA3Qhdh8fAF44AHgAKpMMLgoAHry6THOI1ElXTUbgBzABo4dS4Aay4IAHcuAD5QnFwAFWQ0FzcPbDwrGi5EePiACnY4crgwe2EAfmU7CAc4ADIMMoqOgG0AaThuOAiURAh8BocAXWVezO1sweH8EXEpOBq4LhQJKDh66p7x-Q6iVsxqXDqc1PTCdgBKOHN4trYO6hh1KC4ZADpfqsbhKEaIElHAAEwAFj07AI7HYRhM8ECD2IxQwXBoWmUFAARhAcRRQmdlOgiDRPLkABIpACyABkAJJcMDqGAAUQUWh48QIt30gW+Z3YAHphXAAHo1OFsYD4YoCs4tVoUDRQag6PpfBV4e7odrlbW4b6q9Uwb4woA). Supposing we depend on calling functionality from that `ref` we might want to preserve it.

```ts
<Props extends Record<string, unknown>, RefType extends Ref<unknown>>(
    props?: Props & … & { ref?: RefType }
): …
```

Here we can see that if we pass a ref to this version then we can get the `ref` functionality of the object passed in. With `Ref<unknown>` though we only know that it _is_ a `ref`, but not to what.

<img width="742" alt="Screen Shot 2022-01-12 at 3 01 28 PM" src="https://user-images.githubusercontent.com/5431237/149213147-4c16049d-e01c-45d0-b001-d0f9e0c5f7c9.png">

By the way, I'm not sure what kinds of `ref` we can take here, but if it's to a component or HTML node (vs. to a function or state value) then I think we can limit this to `RefObject` instead of `Ref` and save a little type-checking on the response.

Reply:
Sorry, but I am not following you. I am going to put here the Gutemberg code I linked beforehand.

```javascript
export function useBlockProps(props = {}, { __unstableIsHtml } = {}) {
  // ...
  const mergedRefs = useMergeRefs([
    props.ref,
    useFocusFirstElement(clientId),
    // ...
  ]);

  return {
    ...wrapperProps,
    ...props,
    ref: mergedRefs,
    // ...
  };
}

export default function useMergeRefs(refs) {
  // ...

  return useCallback((value) => {
    // ...
  }, []);
}
```

As you can see, the type of input `ref` is partially (see below) unrelated to output `ref`. You are explicitly trying to use `current` on the output `ref`, but this is always a [callback-ref](https://reactjs.org/docs/refs-and-the-dom.html#callback-refs) in the form of `(value: T) => void`.

In any case you made me realize that, as you noted, [RefObject](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/64d939eb39dc7b80b27da1a63774cfca755670f9/types/react/v16/index.d.ts#L84-L86) (and [MutableRefObject](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/64d939eb39dc7b80b27da1a63774cfca755670f9/types/react/v16/index.d.ts#L901-L903)) is different from [Ref](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/64d939eb39dc7b80b27da1a63774cfca755670f9/types/react/v16/index.d.ts#L88). And the interesting point is that [RefCallback](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/64d939eb39dc7b80b27da1a63774cfca755670f9/types/react/v16/index.d.ts#L87), which is one of the union types of `Ref`, is the type of the callback-ref mentioned above.

So, a reiteration of the stripped code for a question.
```javascript
const mergedRefs = useMergeRefs([
  props.ref,
  useFocusFirstElement(clientId),
  // ...
]);

export function useFocusFirstElement(clientId) {
  const ref = useRef();
  // ...

  useEffect(() => {
    if (!ref.current) {
      return;
    }

    const { ownerDocument } = ref.current;

    // ...
  }, [/* ... */]);

  return ref;
}

export default function useMergeRefs(refs) {
  const element = useRef();
  const previousRefs = useRef([]);
  const currentRefs = useRef(refs);

  return useCallback((value) => {
    assignRef(element, value);
    const refsToAssign = value ? currentRefs.current : previousRefs.current;
    for (const ref of refsToAssign) {
      assignRef(ref, value);
    }
  }, []);
}
```
Given that:
- `value` is _ref-assigned_ to each ref in `refs`
- `refs` contains both our input `ref` and the result of `useFocusFirstElement` (and other hooks)
- `useFocustFirstElement` assumes that `ref.current` is _falsy_ or contains `ownedDocument`
- `ownedDocument` is present in `Node`

Do you think that we can safely assume that `value: Node` and therefore the following?
```typescript
export interface Merged<T extends Node> {
    className: string;
    style: Record<string, unknown>;
    ref: RefCallback<T>;
}

export interface UseBlockProps {
    <Props extends Record<string, unknown>, RefT extends Node>(
        props?: Props & {
            [K in keyof Props]: K extends keyof Reserved ? never : Props[K];
        } & { ref?: Ref<RefT> },
    ): Pick<Props, Exclude<keyof Props, 'ref'>> & Merged<RefT> & Reserved;
}
```
(bikeshedding the name of `RefT`, probably can be just `T`). Notice that this does not change the fact that we should allow `Ref` and not only `RefObject` (because it is correctly handled from `assignRef`), but it should be correct to use `RefCallback` as output `ref` and, at the same time, to correlate the type between input and output refs.

What do you think?

Reply:
very good! thanks for persisting on this, as I missed that the entire `mergedRefs` is the output value for `ref`

I'll try and get back to this ASAP but likely not before the weekend. In this case yeah we should disconnect the input from output and I'll see if I can reproduce and figure out the issue with `Omit` and `Pick`.

Now that I see this, we _could_ expand that so that all the knowable functions on the `ref` are accessible through the types, though I would hate to stall this PR in order to get those in. Probably `RefCallback<unknown>` is good enough for this update.

Reply:
> very good! thanks for persisting on this, as I missed that the entire mergedRefs is the output value for ref

Thank you too, I would have not discovered these subtle details without your review :heart: 

> I'll try and get back to this ASAP but likely not before the weekend. In this case yeah we should disconnect the input from output and I'll see if I can reproduce and figure out the issue with Omit and Pick.

Don't worry, no hurry. I can write what I understood about the problem. First of all, I unfortunately discovered that `dtslint` performs a string serialization from the type, and then it performs an equality comparison between this and what is written after `$ExpectType`. This is pretty clear if you try to use single quotes instead of double quotes -- right, totally different types! :sweat:. But, even worse, using type aliases is a nightmare -- you cannot test anything except for the "name" of the alias, which is pointless.
Now, the issue with `Omit' and `Pick`: it looks like that what I just wrote is not completely true for some types. In particular, `Omit<T, K>` is transformed into `Pick<T, ???>` before type serialization... at a certain point. What I mean is that it takes some time before the error comes out, and if you put `Pick<T, ???>` in the `$ExpectType` (leaving `Omit` in the code), you get the opposite error in much less time. This makes me think that there are two serialization steps inside `dtslint` (which could be related to supporting multiple TS versions?), but the issue is pretty annoying.

> Now that I see this, we could expand that so that all the knowable functions on the ref are accessible through the types, though I would hate to stall this PR in order to get those in. Probably RefCallback<unknown> is good enough for this update.

Do you mean the result types in the array passed to `useMergeRef`? To be honest, the best thing could be to create the types for all of these, then for `useMergeRef` and at that point we would have the real chain of types. But I have to say that this is not exactly a fast thing to do :sweat_smile:. Take your time to think about using `RefCallback<unknown>` or `RefCallback<T>`.
My 2 cents: even if I hate when something returns `unknown` (`any` is even worse), without any strong typing is virtually impossible to use `useBlockProps` for other than spreading into a React element (because it's the only thing showed in the docs AFAIK), therefore using `RefCallback<unknown>` would still improve a bit the current situation respect to a generic `Record<string, unknown>`. Moreover, `RefCallback<unknown>` is forward compatible with, let's say, `RefCallback<Node>`, the opposite is a breaking change. So, I think that you are right about going with `RefCallback<unknown>` :blush:

Reply:
> dtslint performs a string serialization from the type

I didn't realize it was `dtslint` complaining. If that's the case it could be one of those times where it's easier to relax the test code rather than make a design compromise on the actual code (or in this case, the types).

> the best thing could be to create the types for all of these, then for useMergeRef and at that point we would have the real chain of types. 

Exactly 👍 

> But I have to say that this is not exactly a fast thing to do 😅

This is why it might be best for a follow-up PR. It's hard enough to work through this one on its own, let alone with tha jumble in the mix. We can focus on _some value_ now and strive for _more value_ later.

> RefCallback<unknown> is forward compatible with, let's say, RefCallback<Node>

Exactly. Our primary goal is not being _wrong_ with the types. It's okay to sacrifice some usefulness if it helps us move forward _if_ that doesn't bite us later on.

Reply:
> I didn't realize it was dtslint complaining. If that's the case it could be one of those times where it's easier to relax the test code rather than make a design compromise on the actual code (or in this case, the types).

Luckily enough, the workaround I used is (or at least it should be) fine in relationship to the types. It is just something no one would ever write :sweat_smile:. But if you think you will have just a bit of time to test some other approaches during the weekend, I am very open to other solutions!